Add additional API tests for EnhancedSecurity

darryl-apple opened the pull request:
#### 3fd7f86c3308116fdce268db0c3920a932df3378
<pre>
Add additional API tests for EnhancedSecurity
<a href="https://bugs.webkit.org/show_bug.cgi?id=300878">https://bugs.webkit.org/show_bug.cgi?id=300878</a>
<a href="https://rdar.apple.com/162764339">rdar://162764339</a>

Reviewed by Per Arne Vollan.

The additional tests cover subframe navigation and window.open for both same origin
and cross origin requests.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _webContentProcessVariantForFrame:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysEnabledAfterSubFrameNavigationRequestDisables)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysEnabledAfterSubFrameNavigationRequestDisablesCrossOrigin)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysDisabledAfterSubFrameNavigationRequestEnabled)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysDisabledAfterSubFrameNavigationRequestEnabledCrossOrigin)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenWithNoopenerFromEnhancedSecurityPage)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenWithOpenerFromEnhancedSecurityPage)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenNoopenerFromEnhancedSecurityInheritsEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenNoopenerFromStandardWithEnhancedSecurityViaDelegate)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenNoopenerFromEnhancedSecurityWithStandardViaDelegate)):
(TestWebKitAPI::TEST(EnhancedSecurity, LockdownModeTakesPrecedenceOverEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityRequestedWhenLockdownModeActive)):

Canonical link: <a href="https://commits.webkit.org/302585@main">https://commits.webkit.org/302585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca5b4f1d3a19323d74b3d7aadfe5679551727c24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129439 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1696 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80857 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/979f2df8-8712-419d-a566-1af1c91820e1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131310 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1627 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1572 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98579 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66474 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/30914b8f-45ef-4373-83ed-bfd5aec571db) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132386 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79231 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a6f04087-64e8-4cfb-bd6f-3b8defa92c62) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1189 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34067 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80093 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139292 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1486 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1425 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107105 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106949 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27261 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1205 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30800 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54158 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1557 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64920 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1375 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1411 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1479 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

gavin-apple commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
```diff
@@ -422,6 +423,337 @@ static bool isJITEnabled(WKWebView *webView)
+
+}
+
+
```
nit: extra line

pvollan commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
```diff
@@ -1131,13 +1131,31 @@ - (NSString *)_webContentProcessVariantForFrame:(_WKFrameHandle *)frameHandle
+    if (hasLockdownMode)
+        return @"lockdown";
+    if (hasEnhancedSecurity)
+        return @"security";
```
Can this code be shared between the platforms?

pvollan commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
```diff
@@ -422,6 +423,534 @@ static bool isJITEnabled(WKWebView *webView)
+
+    EXPECT_EQ(true, isEnhancedSecurityEnabled(webView.get()));
+#if USE(APPLE_INTERNAL_SDK)
+    EXPECT_STREQ("security", [webView _webContentProcessVariantForFrame:nil].UTF8String);
```
Should we add a comment about why these checks are only enabled with the internal SDK?

pvollan commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
```diff
@@ -422,6 +423,534 @@ static bool isJITEnabled(WKWebView *webView)
+    EXPECT_STREQ("standard", [webView _webContentProcessVariantForFrame:nil].UTF8String);
+    EXPECT_STREQ("standard", [webView _webContentProcessVariantForFrame:[webView firstChildFrame]._handle].UTF8String);
+#endif
+
```
Nit: remove extra line.

darryl-apple committed:
Add additional API tests for EnhancedSecurity
https://bugs.webkit.org/show_bug.cgi?id=300878
rdar://162764339

Reviewed by Per Arne Vollan.

The additional tests cover subframe navigation and window.open for both same origin
and cross origin requests.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _webContentProcessVariantForFrame:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysEnabledAfterSubFrameNavigationRequestDisables)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysEnabledAfterSubFrameNavigationRequestDisablesCrossOrigin)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysDisabledAfterSubFrameNavigationRequestEnabled)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityNavigationStaysDisabledAfterSubFrameNavigationRequestEnabledCrossOrigin)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenWithNoopenerFromEnhancedSecurityPage)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenWithOpenerFromEnhancedSecurityPage)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenNoopenerFromEnhancedSecurityInheritsEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenNoopenerFromStandardWithEnhancedSecurityViaDelegate)):
(TestWebKitAPI::TEST(EnhancedSecurity, WindowOpenNoopenerFromEnhancedSecurityWithStandardViaDelegate)):
(TestWebKitAPI::TEST(EnhancedSecurity, LockdownModeTakesPrecedenceOverEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityRequestedWhenLockdownModeActive)):

Canonical link: https://commits.webkit.org/302585@main
